internal/service/users_test.go

```diff
@@ -214,3 +217,255 @@ func TestUsersService_UpdateUserInfo(t *testing.T) {
+				Email:       "JognDoe@example.com",
+				Password:    "John*123#",
+				FirstName:   "John",
+				LastName:    "",
+				DisplayName: "JohnnyD",
+			},
+			setupMocks: func(ctx context.Context, mockUsers *repoMocks.MockUsers) {},
+			checkError: func(t *testing.T, err error) {
+				var errs validation.Errors
+				ok := errors.As(err, &errs)
+				require.True(t, ok)
+				assert.Equal(t, 1, len(errs))
+				_, ok = errs["last_name"]
+				assert.True(t, ok)
+			},
+		},
+		"validation_weak_password": {
+			email: "JognDoe@example.com",
+			input: &SignupUserInput{
+				Email:       "JognDoe@example.com",
+				Password:    "pw*123",
+				FirstName:   "John",
+				LastName:    "Doe",
+				DisplayName: "JohnnyD",
+			},
+			setupMocks: func(ctx context.Context, mockUsers *repoMocks.MockUsers) {},
+			checkError: func(t *testing.T, err error) {
+				var errs validation.Errors
+				ok := errors.As(err, &errs)
+				require.True(t, ok)
+				assert.Equal(t, 1, len(errs))
+				_, ok = errs["password"]
+				assert.True(t, ok)
+			},
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			mockCtrl := gomock.NewController(t)
+			mockUsers := repoMocks.NewMockUsers(mockCtrl)
+			gen, err := idgen.New(1)
+			assert.NoError(t, err)
+			s := newUsersService(mockUsers, gen)
+			ctx := context.Background()
+			tc.setupMocks(ctx, mockUsers)
+
+			err = s.Signup(ctx, tc.input)
+
+			tc.checkError(t, err)
+		})
+	}
+}
+
+func TestUsersService_Login(t *testing.T) {
+	cases := map[string]struct {
+		email      string
+		input      *LoginInput
+		setupMocks func(context.Context, *repoMocks.MockUsers)
+		checkError func(*testing.T, error)
+	}{
+		"success": {
+			email: "JognDoe@example.com",
+			input: &LoginInput{
+				Email:      "JognDoe@example.com",
+				Password:   password,
+				Persistent: true,
+			},
+			setupMocks: func(ctx context.Context, mockUsers *repoMocks.MockUsers) {
+				hashedPassword, _ := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
+				loggedUser := &core.User{
+					Id:               "1111111",
+					Email:            "doe.h@example.com",
+					FirstName:        "John",
+					LastName:         "Doe",
+					DisplayName:      "JohnnyD",
+					RegistrationDate: someDate,
+					HashedPassword:   hashedPassword,
+					Roles:            []security.Role{security.Student},
+				}
+				mockUsers.EXPECT().GetByEmail(ctx, "JognDoe@example.com").Return(loggedUser, nil).Times(1)
+			},
+			checkError: noError,
+		},
+		"invalid_credentials": {
+			email: "JognDoe@example.com",
+			input: &LoginInput{
+				Email:      "JognDoe@example.com",
+				Password:   password,
+				Persistent: true,
+			},
+			setupMocks: func(ctx context.Context, mockUsers *repoMocks.MockUsers) {
+				loggedUser := &core.User{
+					Id:               "1111111",
+					Email:            "doe.h@example.com",
+					FirstName:        "John",
+					LastName:         "Doe",
+					DisplayName:      "JohnnyD",
+					RegistrationDate: someDate,
```

Comment:
I'd recommend to be more explicit here: hash an explicitly different password (say, password + "qwerty") using bcrypt and put it here. Currently you are testing a somewhat different situation: passing invalid hashed data to bcrypt (if you look closely at the HashedPassword generated by bcrypt, you will see that it has a very specific data format). This is not what we expect when a user enters a wrong password

Reply:
Addressed in the latest commit.